### PR TITLE
Calendar Functionality for Mentors and Mentees

### PR DESCRIPTION
**Pull request by Domm6**
Enable Scheduling of 30-Minute Meetings:
- Mentors and mentees can now schedule 30-minute meetings directly through the calendar interface.

Conflict Handling:
- Added functionality to detect and handle scheduling conflicts to ensure no overlapping meetings.

Display Meetings in Calendar:
- The calendar now visually displays scheduled meetings for both mentors and mentees.
- Improved UI to clearly distinguish between mentor and mentee meetings.

Timezone Handling:
- Fixed bug related to time zone messing up meeting times.
- Implemented unified logic to fetch data for both mentors and mentees based on the user role.
- Refactored code to ensure scalability and maintainability.

Merge Main Branch:
- Merged the latest changes from the main branch to ensure up-to-date functionality and resolve any potential conflicts.

Comprehensive Testing:
- Thoroughly tested the calendar functionalities to ensure smooth operation for both mentor and mentee roles.

Images:
Screenshot 2024-07-08 at 10 19 39 AM
Screenshot 2024-07-08 at 10 19 49 AM
Screenshot 2024-07-08 at 10 20 10 AM

### CODE REVIEW

File: Backend/Routes/users.js [hunk 1]
```diff
@@ -414,4 +415,117 @@ router.get('/connect-requests/mentee/:menteeId', async (req, res) => {
...
+  } catch (error) {
+    res.status(500).json({ error: 'Error checking for conflicts' });
+  }
+});
+
+// create new meeting
+router.post('/meetings', async (req, res) => {
+  const { mentorId, menteeId, scheduledTime, topic } = req.body;
+  const startTime = new Date(scheduledTime).toISOString();
+  const endTime = new Date(new Date(scheduledTime).getTime() + 30 * 60 * 1000).toISOString(); // 30 minute meeting
```
Comment (wspalding) [inline Backend/Routes/users.js]: const these values

File: Frontend/src/components/calendar/CalendarModal.jsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            const response = await fetch(url);
+            if (!response.ok) {
+                throw new Error('Failed to fetch requests');
+            }
+            const data = await response.json();
+            const acceptedRequests = data.requests.filter(request => request.status === 'accepted');
+            setMenteesOrMentors(acceptedRequests.map(request => ({
+                id: user.userRole === 'mentor' ? request.menteeId : request.mentorId,
+                name: user.userRole === 'mentor' ? request.menteeName : request.mentorName
+            })));
```
Comment (wspalding) [inline Frontend/src/components/calendar/CalendarModal.jsx]: Can a single user be both a mentor and a mentee? or would that necessarily require 2 accounts?

File: Frontend/src/components/calendar/CalendarModal.jsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                const url = user.userRole === 'mentor'
+                    ? `${config.apiBaseUrl}/mentors/${user.id}`
+                    : `${config.apiBaseUrl}/mentees/${user.id}`;
+
+                const response = await fetch(url);
+                if (!response.ok) {
+                    throw new Error('Failed to fetch user data');
+                }
+                const data = await response.json();
+                setUserData(user.userRole === 'mentor' ? data.mentor : data.mentee);
```
Comment (wspalding) [inline Frontend/src/components/calendar/CalendarModal.jsx]: Looks like checking if a user is a mentor or mentee is very common. Would make sense to encapsulate that logic into a helper function

File: Backend/Routes/users.js (hunk 1 above)
Comment (sbrina813) [reply Backend/Routes/users.js]: like const MS_IN_HALF_HOUR =

File: Backend/Routes/users.js [hunk 4]
```diff
@@ -414,4 +415,117 @@ router.get('/connect-requests/mentee/:menteeId', async (req, res) => {
...
+
+  try {
+    // Check if there is a conflict
+    const conflict = await Meeting.findOne({
```
Comment (sbrina813) [inline Backend/Routes/users.js:495]: what about creating a util for the conflict?

File: Frontend/src/components/calendar/CalendarModal.jsx (hunk 2 above)
Comment (Domm6) [reply Frontend/src/components/calendar/CalendarModal.jsx]: A single user can be only a mentor or a mentee, so it requires 2 separate accounts.

File: Backend/Routes/users.js (hunk 1 above)
Comment (Domm6) [reply Backend/Routes/users.js]: Resolved

File: Frontend/src/components/calendar/CalendarModal.jsx (hunk 3 above)
Comment (Domm6) [reply Frontend/src/components/calendar/CalendarModal.jsx]: Resolved

File: Backend/Routes/users.js (hunk 4 above)
Comment (Domm6) [reply Backend/Routes/users.js:495]: I plan on making a util file to refactor my code after I complete my technical challenges. Thanks for the heads up as I will keep this in mind to add to it.

**Review by jzwang43 — APPROVED** [review]
Nice integration of the full calendar package, thanks for addressing the feedback as well

**Pull request merged**